core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala

```diff
@@ -0,0 +1,179 @@
+import akka.http.scaladsl.model.headers.OAuth2BearerToken
+import bio.terra.workspace.model.CreateCloudContextResult
+import bio.terra.workspace.model.JobReport.StatusEnum
+import com.typesafe.scalalogging.LazyLogging
+import io.opencensus.scala.Tracing.traceWithParent
+import io.opencensus.trace.Span
+import org.broadinstitute.dsde.rawls.config.MultiCloudWorkspaceConfig
+import org.broadinstitute.dsde.rawls.dataaccess.slick.{DataAccess, ReadWriteAction}
+import org.broadinstitute.dsde.rawls.dataaccess.workspacemanager.WorkspaceManagerDAO
+import org.broadinstitute.dsde.rawls.dataaccess.{AttributeTempTableType, SlickDataSource}
+import org.broadinstitute.dsde.rawls.model.{AttributeName, ErrorReport, GoogleProjectId, MultiCloudWorkspaceRequest, UserInfo, Workspace, WorkspaceAttributeSpecs, WorkspaceShardStates, WorkspaceType, WorkspaceVersions}
+import org.broadinstitute.dsde.rawls.util.OpenCensusDBIOUtils.traceDBIOWithParent
+import org.broadinstitute.dsde.rawls.util.Retry
+import org.broadinstitute.dsde.rawls.{RawlsException, RawlsExceptionWithErrorReport}
+import org.joda.time.DateTime
+import slick.dbio.DBIO
+import slick.jdbc.TransactionIsolation
+
+import java.util.UUID
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+
+object MultiCloudWorkspaceService {
+  def constructor(dataSource: SlickDataSource,
+                  workspaceManagerDAO: WorkspaceManagerDAO,
+                  multiCloudWorkspaceConfig: MultiCloudWorkspaceConfig)
+                 (userInfo: UserInfo)
+                 (implicit ec: ExecutionContext, system: ActorSystem): MultiCloudWorkspaceService = {
+    new MultiCloudWorkspaceService(
+      userInfo,
+      workspaceManagerDAO,
+      multiCloudWorkspaceConfig,
+      dataSource
+    )
+  }
+}
+
+
+/**
+  * This service knows how to provision a new "multi-cloud" workspace, a workspace managed by terra-workspace-manager.
+  */
+class MultiCloudWorkspaceService(userInfo: UserInfo,
+                                 workspaceManagerDAO: WorkspaceManagerDAO,
+                                 multiCloudWorkspaceConfig: MultiCloudWorkspaceConfig,
+                                 dataSource: SlickDataSource)
+                                (implicit ec: ExecutionContext, val system: ActorSystem) extends LazyLogging with Retry {
+
+  def createMultiCloudWorkspace(workspaceRequest: MultiCloudWorkspaceRequest, parentSpan: Span = null) = {
+    if (!multiCloudWorkspaceConfig.multiCloudWorkspacesEnabled) {
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotImplemented, "MC workspaces are not enabled"))
+    }
+
+    for {
+      workspace <- traceWithParent("withNewMcWorkspaceContext", parentSpan)(
+        opSpan => dataSource.inTransactionWithAttrTempTable(Set(AttributeTempTableType.Workspace))({ dataAccess =>
+          createWorkspace(workspaceRequest, dataAccess, opSpan) { createdWorkspace =>
```

Comment:
Per our discussion today, does there need to be cleanup on deletion?

Reply:
I expect that will be the case when we implement WOR-80 (building the deletion endpoint). As of now, this endpoint isn't allocating any resources other than a workspace + cloud context record in WSM.